### DISCUSSION TITLE
Firefox GHOSTS support

### DISCUSSION THREAD

**Discussion by rbreesems** [cmu-sei/GHOSTS#152] [General]
Supporting both Chrome and Firefox in GHOSTS is important to our event guys, and I have been working testing GHOSTS Firefox in our event environment, particularly the 'random' command (Chrome works great in all tests).

Since Firefox does not have a command line option like Chrome for accepting certs, I have added logic that navigates the Firefox security bypass pop-ups when an untrusted cert exception is encountered. However, I ran into another problem that has been driving me buggy.  A lot of our pages have links to downloadable files. With  Selenium Chrome, this is no problem, the file associated with the link is downloaded and browsing continues. However, with Selenium Firefox, the call to 
```
BaseBrowserHandler.cs/MakeRequest/Driver.Navigate().GoToUrl(config.Uri);
```

 succeeds (the file is downloaded) but then an exception is thrown that looks like:

```
BaseBrowserHandler.MakeRequest|The HTTP request to the remote WebDriver server for URL http://localhost:someport>/session/<someUUid>/url timed out after 60
```

Web searches show that this has been a common problem with Selenium for several years, especially with Firefox, with conflicting info on how to solve it, including some info that it is not solvable.  I have tried several different combinations of Firefox/geckodriver version combinations, and tried several different suggestions for fixing this.  I have also tried updating to the latest Selenium Webdriver.  Nothing solves the problem.

My best guess is that it is an inherent problem with `Navigate().GoToUrl()`  for downloadable file links with Firefox.  On an experimental basis, I have replaced this with Javascript navigation to the actual element on the page (instead of the link), and this works fine (download works and browsing continues).  However, using elements instead of links bypasses all of the link management/selection that is done for 'random'.  I have spent a considerable amount of time on this problem and need to move on.

If you have a solution, let me know. Else, I am about to spinup a new branch that  will do things differently for Firefox. In order to keep upward combability, my thought is to make a new Handler class, (perhaps `BrowserFirefoxRobust`) that inherits from `BrowserFirefox`, and then make things like `MakeRequest` a virtual method that can be overridden. This would keep all of the same behavior for `BrowserFireFox`, `BrowserChrome`, but `BrowserFirefoxRobust` would handle links and browsing much differently.

**Comment by sei-dupdyke** [comment]
I have a note to myself to investigate much of the [discussion here](https://news.ycombinator.com/item?id=33537772) further. Maybe there are other options to selenium that would be easier to use and more robust?

**Reply by rbreesems** [reply]
Thanks for the response.

I just finished browsing that discussion.  Trying to start with something other than Selenium at this stage has too many risks (how buggy would it be, how long would it take to come up to speed on it, etc) . Also, a lot of the advantages of the other platforms seemed to be speed-related for automated testing in terms of browser actions, which is not an issue for us. 

I believe that I will just continue on with the new Handler (`BrowserFirefoxRobust`) that will isolate my changes and keep everything upward compatible with any current timelines, current handlers.
